### PR TITLE
feature(cp): implement archive & -L

### PR DESCRIPTION
**Pull request by sylvestre**
Real life test case:

```
set -e

rm -rf dir dir-copy-archive-rust
mkdir dir
echo "a" > dir/1
echo "b" > dir/2
touch -d "2 hours ago" dir/1
chmod 705 dir/*
cd dir
ln -s 1 1.link
ln -s 2 2.link
cd -
cp --archive dir dir-copy-archive
ls -al dir-copy-archive|grep "rwx---r-x"
ls -al dir-copy-archive|grep "1.link -> 1"
./target/debug/coreutils cp --archive -v dir dir-copy-archive-rust
ls -al dir-copy-archive-rust|grep "rwx---r-x"
ls -al dir-copy-archive-rust|grep "1.link -> 1"
ls -al dir-copy-archive-rust|grep "2.link -> 2"
```

### CODE REVIEW

File: cp-arch-test.sh [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+cd dir
+ln -s 1 1.link
+ln -s 2 2.link
+cd -
+cp --archive dir dir-copy-archive
+ls -al dir-copy-archive|grep "rwx---r-x"
+ls -al dir-copy-archive|grep "1.link -> 1"
+./target/debug/coreutils cp --archive -v dir dir-copy-archive-rust
+ls -al dir-copy-archive-rust|grep "rwx---r-x"
+# les symlinks ne sont pas créés
```
Comment (nbraud) [inline cp-arch-test.sh:17]: This probably shouldn't be in French, and also I'm confused by the meaning of the comment here.

File: src/uu/cp/src/cp.rs [hunk 2]
```diff
@@ -1223,6 +1259,7 @@ fn copy_helper(source: &Path, dest: &Path, options: &Options) -> CopyResult<()>
         } else {
             dest.into()
         };
+        // le pb est que le symlink doit toujours etre cree
```
Comment (sylvestre) [inline src/uu/cp/src/cp.rs:1262]:
```suggestion

```

File: src/uu/cp/src/cp.rs [hunk 3]
```diff
@@ -1211,6 +1246,7 @@ fn copy_helper(source: &Path, dest: &Path, options: &Options) -> CopyResult<()>
     } else if options.no_dereference && fs::symlink_metadata(&source)?.file_type().is_symlink() {
         // Here, we will copy the symlink itself (actually, just recreate it)
         let link = fs::read_link(&source)?;
+
```
Comment (sylvestre) [inline src/uu/cp/src/cp.rs:1249]:
```suggestion
```

**Comment by sylvestre** [comment]
I am planning to land this Thursday 5st if no objection!

**Comment by rivy** [comment]
I haven't really had the time to review this... (busy with COVID uptick at the hospitals).
There are a few style notes/warnings and several CodeCov comments about new, none-test-covered, code.
And I haven't had the chance to actually run/test the new option.
I should be able to get to it over next weekend if that's not too long.

File: src/uu/cp/src/cp.rs [hunk 4]
```diff
@@ -873,7 +888,7 @@ fn copy_source(
 }
 
 #[cfg(target_os = "windows")]
-fn adjust_canonicalization(p: &Path) -> Cow<Path> {
+fn adjust_canonicalization<'a>(p: &'a Path) -> Cow<'a, Path> {
```
Comment (rivy) [inline src/uu/cp/src/cp.rs:891]:
```suggestion
fn adjust_canonicalization(p: &Path) -> Cow<Path> {
```

`cargo clippy` says the explicit lifetimes are unneeded. My test compile `cargo +1.32.0 test` compiled and tested correctly.

File: tests/by-util/test_cp.rs [hunk 5]
```diff
@@ -491,3 +499,220 @@ fn test_cp_no_deref_folder_to_folder() {
     let path_to_check = path_to_new_symlink.to_str().unwrap();
     assert_eq!(at.read(&path_to_check), "Hello, World!\n");
 }
+
```
Comment (rivy) [inline tests/by-util/test_cp.rs:502]: So, all the tests target only "linux"? ☹️

We definitely need to be testing all platforms. If there is some difficulty in getting them to work on MacOS/Windows, I can offer some time to help at a later date.

And, re-reading [the docs](https://doc.rust-lang.org/reference/conditional-compilation.html#target_os), I think `target_os = "unix"` is incorrect; it should be `target_os = "linux"`.

"unix" is a `target_family`.

**Review by rivy — CHANGES_REQUESTED** [review]
Minor syntax change request, but I'm concerned that the tests are all "linux" gated.

File: tests/by-util/test_cp.rs (hunk 5 above)
Comment (sylvestre) [reply tests/by-util/test_cp.rs:502]: Yeah, I don't know how to fix these :/

Symlinks are tricky on Windows

**Comment by sylvestre** [comment]
Sorry, I know this isn't perfect but this work and it is improving the overall code coverage.
Landing as it has been around since August

**Pull request merged**